pubsub/gossipsub/gossipsub-v2.0.md

```diff
@@ -0,0 +1,80 @@
+# Episub - Gossipsub Extension
+
+
+# Overview
+
+This document aims to provide a minimal extension to the [gossipsub v1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md) protocol, that supersedes the previous [episub](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/episub.md) proposal.
+
+The proposed extensions are backwards-compatible and aim to enhance the efficiency (minimize amplification/duplicates and decrease message latency) of the gossip mesh networks by dynamically adjusting the messages sent by mesh peers based on a local view of message duplication and latency. 
+
+In more specific terms, two new control messages are introduced, `CHOKE` and `UNCHOKE`. When a Gossipsub router is receiving many duplicates on a particular mesh, it can send a `CHOKE` message to it's mesh peers that are sending duplicates slower than its fellow mesh peers. Upon receiving a `CHOKE` message, a peer is informed to no longer propagate mesh messages to the sender of the `CHOKE` message, rather lazily (in every heartbeat) send it's gossip. 
+
+A Gossipsub router may notice that it is receiving messages via gossip from it's `CHOKE`'d peers faster than it receives them via the mesh. In this case it may send an `UNCHOKE` message to the peer to inform the peer to resume propagating messages in the mesh. 
+The router may also notice that it is receiving messages from gossip from peers not in the mesh (fanout) faster than it receives messages in the mesh. It may then add these peers into the mesh. 
+
+The modifications outlined above intend to optimize the Gossipsub mesh to receive minimal duplicates from peers with the lowest latency.
+
+# Specification
+
+## Protocol Id
+
+Nodes that support this Gossipsub extension should additionally advertise the version number `2.0.0`. Gossipsub nodes can advertise their own protocol-id prefix, by default this is `meshsub` giving the default protocol id:
+- `/meshsub/2.0.0`
+
+## Parameters
+
+This section lists the configuration parameters that control the behaviour of the mesh optimisation.
+
+
+| Parameter | Description  | Reasonable Default |
+| -------- | -------- | -------- |
+| `D_non_choke`     | The minimum number of peers in a mesh that must remain unchoked. | `D_lo` |
+| `choke_heartbeat_interval` | The number of heartbeats before assessing and applying `CHOKE`/`UNCHOKE` control messages and adding `D_max_add`. | 20 |
+| `D_max_add` | The maximum number of peers to add into the mesh (from fanout) if they are performing well per `choke_heartbeat_interval`. | 1 |
```

Comment:
I am not sure we need this, at least when opportunistic grafting is enabled.
Those peer will have a good score and get picked up.
We should analyze this, and have a mention here.

Reply:
I had a few thoughts on this. 
1. I was working under the assumption that gossipsub scoring is optional (it still is in rust). I was hoping to make this extension independent of scoring, i.e users that don't enable or want scoring can still benefit from these additions.
2. I know this is an issue with our scoring params, (in my case and potentially others) the scoring params are not strict enough to discern mesh peers (they are all positive) so we don't do opportunistic grafting that often.

Reply:
ok, this makes sense.
Lets look at some experimental data first.